src/resources.ts

```diff
@@ -65,31 +69,61 @@ export function useUser(id: string) {
 export function useCountries() {
   const countryFetcher = useCallback(
     (options: AuthenticatedHttpOptions) => fetchCountries(options),
-    [],
+    []
   );
   const { loading, resource: countries } = useAuthenticatedHttpResource([], countryFetcher);
   return { countries, loading };
 }
 
+const SHARING_CODE_TIMER_INTERVAL = 500;
+
+function getTimeFromNow(date: Date) {
+  return (date.getTime() - new Date().getTime()) / 1000;
+}
+
 export function useSharingCode(userId: string) {
   const codeCreator = useCallback(
     (options: AuthenticatedHttpOptions) => createSharingCodeForUserId(userId, options),
-    [userId],
+    [userId]
   );
-  const { loading, resource: sharingCode } = useAuthenticatedHttpResource(null, codeCreator);
-  return { loading, sharingCode };
+  const {
+    loading,
+    resource: sharingCode,
+    reloadResource: reloadSharingCode,
+  } = useAuthenticatedHttpResource(null, codeCreator);
+  const [timeUntilExpiry, setTimeUntilExpiry] = useState(0);
+  const [originalTimeUntilExpiry, setOriginalTimeUntilExpiry] = useState(0);
+  useEffect(() => {
+    if (sharingCode) {
+      const cancelToken = http.CancelToken.source();
+      const expiryDate = new Date(sharingCode.expiryTime);
+      setOriginalTimeUntilExpiry(getTimeFromNow(expiryDate));
+      const interval = setInterval(() => {
+        const timeLeft = getTimeFromNow(expiryDate);
+        setTimeUntilExpiry(timeLeft);
+        if (timeLeft <= 3) {
+          reloadSharingCode(cancelToken.token);
+        }
+      }, SHARING_CODE_TIMER_INTERVAL);
+      return () => {
+        clearInterval(interval);
+        cancelToken.cancel();
+      };
+    }
+  }, [reloadSharingCode, sharingCode]);
+  return { loading, sharingCode, timeUntilExpiry, originalTimeUntilExpiry };
```

Comment:
Why do we need to expose those outside the hook?

Reply:
When we want to expose a good-looking progress spinner (how much time there is left until expire). I started doing it with this pr but decided it might be a little more confusing than not having it since the scan process is very fast. Decided to wait for design, felt like keeping this around till then.